shell/platform/darwin/ios/framework/Source/FlutterAppDelegate.mm

```diff
@@ -149,18 +155,29 @@ - (BOOL)openURL:(NSURL*)url {
                      if (didTimeout) {
```

Comment:
I remembered this was added to workaround the issue that when we receive openurl, we don't know what is the current state of flutter. If we send if before the first frame is drawn, the message will be dropped. It is also too late to set initialroute at this point since it is already consumed by the framework.

Reply:
but this was a long time ago, I am not sure if this is still the case

Reply:
It seems like `FlutterAppDelegate` needs to know a lot of details about the engine, the navigationChannel, and the available method.  What if this changed to:
```objc
- (void)openURL:(NSURL*)url
              options:(NSDictionary<UIApplicationOpenExternalURLOptionsKey, id>*)options
    completionHandler:(void (^)(BOOL success))completion {
  if (![self isFlutterDeepLinkingEnabled]) {
    completion(NO);
  } else {
    FlutterViewController* flutterViewController = [self rootFlutterViewController];
    if (flutterViewController) {
      [flutterViewController openDeepLink:url completionHandler:completion];
    } else {
      FML_LOG(ERROR) << "Attempting to open an URL without a Flutter RootViewController.";
      completion(NO);
    }
  }
}
```
FlutterViewController_Internal.h
```objc
- (void)openDeepLink:(NSURL*)url completionHandler:(void (^)(BOOL success))completion;
```
and then in FlutterViewController:
```objc
- (void)openDeepLink:(NSURL*)url completionHandler:(void (^)(BOOL success))completion {
  [_engine.get()
      waitForFirstFrame:3.0
               callback:^(BOOL didTimeout) {
                 if (didTimeout) {
                   FML_LOG(ERROR) << "Timeout waiting for the first frame when launching an URL.";
                   completion(NO);
                 } else {
                   // invove the method and get the result
                   [[_engine.get() navigationChannel]
                       invokeMethod:@"pushRouteInformation"
                          arguments:@{
                            @"location" : url.absoluteString ?: [NSNull null],
                          }
                             result:^(id _Nullable result) {
                               BOOL success =
                                   [result isKindOfClass:[NSNumber class]] && [result boolValue];
                               if (!success) {
                                 // Logging the error if the result is not successful
                                 FML_LOG(ERROR) << "Failed to handle route information in Flutter.";
                               }
                               completion(success);
                             }];
                 }
               }];
}
```

Reply:
yup, It does seem like the viewcontroller may be a better place to put these logic

Reply:
thanks, this looks good, updated!